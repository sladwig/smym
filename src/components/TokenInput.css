.token-input {
    border: 1px solid gold;
    display: flex;
    flex-direction: row;
    height: 20px;
}
.display {
    border: 1px solid blue;
cursor: text;
}
.post-display {
    flex-grow: 1;
}
.input-field {
    border: 1px solid red;
    border-right: 0px solid transparent;
    border-left: 0px solid transparent;
    text-align: left;
    padding: 0 .125em;
}
.input-field:focus {
    outline: 0px solid transparent;
}

.display {
    border-right: 0px solid transparent;
    border-left: 0px solid transparent;
    padding: 0 .125em;

}