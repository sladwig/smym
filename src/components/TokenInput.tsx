import React, { useRef, useCallback, useState, useEffect, useMemo } from 'react';
import './TokenInput.css';
import { observer } from 'mobx-react-lite';
import { TokenDisplay, tokenInputFocus } from './TokenDisplay';
import create from 'zustand';
import classnames from 'classnames';
import { SearchIcon, CancelIcon } from './TransactionInput';
import { ReactComponent as SearchIconSvg } from '../images/search-icon.svg';
import { SuggestionBox, useSuggestionStore } from './SuggestionBox';
import { MovingEye } from './MovingEye';

const initialValue = { value: '', position: 0, hasFocus: false, external: false };
const [useInputStore, inputStore] = create(set => ({
    ...initialValue,
    reset: () => set(initialValue),
}));

const { setState } = inputStore;
export { useInputStore, inputStore, setState };

const external = false;
export const TokenInput = observer(() => {
    const inputRef = useRef<HTMLInputElement>(null);

    const [value, position, hasFocus, externalUpdate] = useInputStore(s => [
        s.value,
        s.position,
        s.hasFocus,
        s.external,
    ]);

    // position update
    const initalizeCaretFromStore = useCallback((pos: number) => {
        inputRef.current!.setSelectionRange(pos, pos);
    }, []);

    useEffect(() => {
        const updatePosition = () => {
            const position = inputRef.current!.selectionStart;
            if (position !== undefined) setState({ position, external });
        };
        document.addEventListener('keydown', updatePosition);
        document.addEventListener('keyup', updatePosition);
        return () => {
            document.removeEventListener('keydown', updatePosition);
            document.removeEventListener('keyup', updatePosition);
        };
    }, []);

    useEffect(() => {
        if (hasFocus) {
            inputRef.current!.focus();
        }
    }, [hasFocus]);

    useEffect(() => {
        if (hasFocus && externalUpdate) {
            initalizeCaretFromStore(position);
        }
    }, [hasFocus, externalUpdate, position]);

    const isDev = useMemo(() => window.localStorage.getItem('isDev'), []);
    return (
        <>
            <TokenDisplay />
            <input
                type="text"
                value={value}
                key={`input-field`}
                ref={inputRef}
                autoFocus={true}
                className={classnames('input-field', { isDev })}
                onFocus={tokenInputFocus}
                onBlur={() => setState({ hasFocus: false })}
                onChange={event => {
                    const position = inputRef.current!.selectionStart;
                    console.log('onchange current pos', event.target.value);
                    setState({ value: event.target.value, position, external });
                }}
            />
        </>
    );
});

export const TransactionInputArea = () => {
    const [active, setActive] = useState(false);
    useEffect(() => {
        if (active) tokenInputFocus();
    }, [active]);

    const mode = useSuggestionStore(state => state.mode);
    const suggestion = mode !== 'none';
    const height = suggestion ? 194 : 104;
    return (
        <div className={classnames('top-wrapper', { active, suggestion })} style={{ height }}>
            <div
                className={classnames('transaction-input-area', 'heading-1')}
                onClick={() => setActive(true)}
            >
                {!active && (
                    <>
                        <MovingEye />
                        <SearchIcon onClick={() => setActive(true)} />
                    </>
                )}
                {active && (
                    <>
                        <SearchIconSvg style={{ marginRight: 29 }} />
                        <TokenInput />
                        <CancelIcon
                            onClick={e => {
                                console.log('ola');
                                setActive(false);
                                inputStore.getState().reset();
                                e.stopPropagation();
                            }}
                            style={{ marginLeft: 29 }}
                        />
                    </>
                )}
            </div>
            <SuggestionBox />
        </div>
    );
};
