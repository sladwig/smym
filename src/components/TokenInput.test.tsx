import { TokenInput } from './TokenInput';
import { render } from '@testing-library/react';
import React, { ReactElement } from 'react';

const renderAndActOn = (el: ReactElement) => {
    const fromRender = render(el);
    const access = {
        tokenInput: fromRender.container.querySelector('.token-input') as HTMLDivElement,
        preDisplay: fromRender.container.querySelector('.pre-display') as HTMLDivElement,
        inputField: fromRender.container.querySelector('.input-field') as HTMLDivElement,
        postDisplay: fromRender.container.querySelector('.post-display') as HTMLDivElement,
        placeholderField: fromRender.container.querySelector('.pre-display') as HTMLDivElement,
    };
    return { ...fromRender, ...access };
};

describe('TokenInput', () => {
    test('renders', () => {
        const { container } = render(<TokenInput />);
        expect(container).toMatchSnapshot();
    });
});
