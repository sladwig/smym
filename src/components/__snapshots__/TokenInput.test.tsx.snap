// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenInput renders 1`] = `
<div>
  <div
    class="token-input"
  >
    <div
      class="display pre-display"
    />
    <div
      class="input-field"
      contenteditable="true"
      spellcheck="false"
      tabindex="0"
    />
    <div
      class="display post-display"
    />
  </div>
</div>
`;
